ssh-agent.c

```diff
@@ -198,16 +201,143 @@ idtab_init(void)
 	idtab->nentries = 0;
 }
 
+static void
+free_dest_constraints(struct dest_constraint *dcs, size_t ndcs)
+{
+	size_t i;
+
+	for (i = 0; i < ndcs; i++) {
+		free(dcs->user);
+		free(dcs->hostname);
+		sshkey_free(dcs->key);
+	}
+	free(dcs);
+}
+
 static void
 free_identity(Identity *id)
 {
 	sshkey_free(id->key);
 	free(id->provider);
 	free(id->comment);
 	free(id->sk_provider);
+	free_dest_constraints(id->dest_constraints, id->ndest_constraints);
 	free(id);
 }
 
+/*
+ * check whether a destination constraint matches a hostkey (attached to
+ * a SocketEntry). Returns 0 on match or -1 otherwise. Matching hostname (if
+ * any) may be returned via *hostnamep.
+ */
+static int
+check_destination_constraint(struct sshkey *hostkey, struct dest_constraint *d,
+    char *user, const char **hostnamep)
+{
+	const char *reason;
+
+	if (hostnamep != NULL)
+		*hostnamep = NULL;
+	if (hostkey == NULL || d->key == NULL)
+		goto matchuser;
+	if (!sshkey_is_cert(hostkey)) {
+		if (!sshkey_equal(hostkey, d->key))
+			return -1;
+		goto matchuser;
+	}
+	/* hostkey is a certificate */
+	if (!d->is_ca)
+		return -1;
+	if (!sshkey_cert_check_host(hostkey, d->hostname, 1,
+	    SSH_ALLOWED_CA_SIGALGS, &reason)) {
+		debug_f("reason: %s", reason);
+		return -1;
+	}
+ matchuser:
+	if (d->user != NULL && user != NULL && !match_pattern(d->user, user))
```

Comment:
should 'user' be required if d->user is set?

Reply:
I don't think so - e.g. when in process_request_identities() a user will never be presented, but we still want to filter as much as possible there.